code/game/turfs/simulated/floor_icon.dm

```diff
@@ -132,28 +140,171 @@ var/list/flooring_cache = list()
 			F.update_icon()
 	update_openspace()
 
-//Tests whether this floor/ing will smooth with the specified turf
-/turf/simulated/floor/proc/test_link(var/turf/T)
-	//is_wall is true for wall turfs and for floors containing a low wall
-	var/is_linked = FALSE
-	if(T.is_wall)
-		if(flooring.wall_smooth)
-			is_linked = TRUE
 
-	//If is_hole is true, then it's space or openspace
-	else if(T.is_hole)
-		if(flooring.space_smooth)
-			is_linked = TRUE
+//Tests whether this flooring will smooth with the specified turf
+//You can override this if you want a flooring to have super special snowflake smoothing behaviour
+/decl/flooring/proc/test_link(var/turf/origin, var/turf/T, var/countercheck = FALSE)
+
+	var/is_linked = FALSE
+	if (!countercheck)
+		//is_wall is true for wall turfs and for floors containing a low wall
+
+		if(T.is_wall)
+			if(wall_smooth)
+				is_linked = TRUE
 
-	//If we get here then its a normal floor
-	else if (istype(T, /turf/simulated/floor))
-		var/turf/simulated/floor/t = T
-		if (flooring.floor_smooth || t.flooring.name == flooring.name)
-			is_linked = TRUE
+
+		//If is_hole is true, then it's space or openspace
+		else if(T.is_hole)
+			if(space_smooth)
+				is_linked = TRUE
+
+
+		//If we get here then its a normal floor
+		else if (istype(T, /turf/simulated/floor))
+			var/turf/simulated/floor/t = T
+			//If the floor is the same as us,then we're linked,
+			if (t.flooring.name == name)
+				/*
+					But there's a caveat. To make atom black/whitelists work correctly, we also need to check that
+					they smooth with us. Ill call this counterchecking for simplicity.
+					This is needed to make both turfs have the correct borders
+
+					To prevent infinite loops we have a countercheck var, which we'll set true
+				*/
+				if (smooth_movable_atom != SMOOTH_NONE)
+					//We do the countercheck, passing countercheck as true
+					is_linked = test_link(T, origin, countercheck = TRUE)
+
+
+			else if (floor_smooth == SMOOTH_ALL)
+				is_linked = TRUE
+
+			else if (floor_smooth != SMOOTH_NONE)
+
+
+
+				//If we get here it must be using a whitelist or blacklist
+				if (floor_smooth == SMOOTH_WHITELIST)
+					for (var/v in flooring_whitelist)
+						if (istype(t.flooring, v))
+							//Found a match on the list
+							is_linked = TRUE
+							break
+				else if(floor_smooth == SMOOTH_BLACKLIST)
+					is_linked = TRUE //Default to true for the blacklist, then make it false if a match comes up
+					for (var/v in flooring_whitelist)
+						if (istype(t.flooring, v))
+							//Found a match on the list
+							is_linked = FALSE
+							break
+	else
+		//If this is a countercheck, we skip all of the above, start off with true, and go straight to the atom lists
+		is_linked = TRUE
```

Comment:
why do that? 
it doesn't really make much difference either way

Reply:
Yeah but that structure will be easy to read and update. 
Structures like that is not good style and here it can be easily avoided:
```
if(condition 1)
	if(condition 2)
		if(condition 3)
			if(condition 4)
				if(condition 5)
					if(condition 6)
						if(condition 7)
							if(condition 8)
								return
							else
								expression 8
						else
							expression 7
					else
						expression 6
				else
					expression 5
			else
				expression 4
		else
			expression 3
	else
		expression 2
else
	expression 1

```